File: docs/examples/cpp/interfaces.cpp
// Interfaces usage examples.

// --------------------------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.
// --------------------------------------------------------------------------------------------

// start-cell-1
#include <qdk/chemistry.hpp>

// Create an SCF solver that uses the QDK/Chemistry library as solver
auto scf = ScfSolverFactory::create();

// Configure it using the standard settings interface
scf->settings().set("basis_set", "cc-pvdz");
scf->settings().set("method", "hf");

// Run calculation with the same API as native implementations
auto [energy, orbitals] = scf->solve(structure);
// end-cell-1

// start-cell-2
#include <iostream>
#include <qdk/chemistry.hpp>

// Get a list of available SCF solver implementations
auto available_solvers = ScfSolverFactory::available();
for (const auto& solver : available_solvers) {
  std::cout << solver << std::endl;
}

// Get documentation for a specific implementation
std::cout << ScfSolverFactory::get_docstring("default") << std::endl;
// end-cell-2

// start-cell-3
#include <qdk/chemistry.hpp>

#include "custom_chemistry_package.hpp"

namespace qdk::chemistry {
namespace algorithms {

class CustomScfSolver : public ScfSolver {
 public:
  CustomScfSolver() = default;

  std::tuple<double, data::Orbitals> solve(
      const data::Structure& structure) override {
    // Convert QDK/Chemistry structure to custom package format
    auto custom_mol = convert_to_custom_format(structure);

    // Run calculation with custom package
    auto result = custom_chemistry::run_scf(
        custom_mol, settings().get<std::string>("basis_set"),
        settings().get<std::string>("method"));

    // Convert results back to QDK/Chemistry format
    double energy = result.energy;
    data::Orbitals orbitals = convert_from_custom_format(result.orbitals);

    return {energy, orbitals};
  }

 private:
  custom_chemistry::Molecule convert_to_custom_format(
      const data::Structure& structure);
  data::Orbitals convert_from_custom_format(
      const custom_chemistry::Orbitals& orbitals);
};

// Register in a static initializer block
namespace {
bool registered = ScfSolverFactory::register_implementation(
    "custom", []() { return std::make_unique<CustomScfSolver>(); },
    "Interface to Custom Chemistry Package");
}  // anonymous namespace

}  // namespace algorithms
}  // namespace qdk::chemistry
// end-cell-3

// start-cell-4
// Set general options that work across all backends
scf->settings().set("basis_set", "cc-pvdz");
scf->settings().set("max_iterations", 100);
// end-cell-4

File: docs/examples/python/interfaces.py
"""Example demonstrating algorithm interface usage."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

import numpy as np
from qdk_chemistry.data import Structure

coords = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.4]])
structure = Structure(coords, ["H", "H"])

# start-cell-1
from qdk_chemistry.algorithms import create

# Create an SCF solver using the factory
scf_solver = create("scf_solver")

# Configure it using the standard settings interface
scf_solver.settings().set("basis_set", "cc-pvdz")
scf_solver.settings().set("method", "hf")

# Run calculation - returns (energy, wavefunction)
energy, wavefunction = scf_solver.run(structure, charge=0, spin_multiplicity=1)
orbitals = wavefunction.get_orbitals()

print(f"SCF Energy: {energy:.10f} Hartree")
# end-cell-1


# start-cell-2
from qdk_chemistry.algorithms import available

# List available implementations for each algorithm type
print("Available SCF solvers:", available("scf_solver"))
print("Available Hamiltonian constructors:", available("hamiltonian_constructor"))
print("Available orbital localizers:", available("orbital_localizer"))
print("Available MC calculators:", available("multi_configuration_calculator"))
# end-cell-2


# start-cell-4
# All algorithms use a consistent settings interface
scf = create("scf_solver")

# Set general options that work across implementations
scf.settings().set("basis_set", "sto-3g")
scf.settings().set("max_iterations", 100)
scf.settings().set("tolerance", 1.0e-8)

# Query available settings for an algorithm
print(f"SCF settings: {scf.settings().keys()}")

# Get a setting value
max_iter = scf.settings().get("max_iterations")
print(f"Max iterations: {max_iter}")
# end-cell-4


